desc/utils_batched_vectorize.py

```diff
@@ -0,0 +1,462 @@
+# netket/jax/_chunk_utils.py
+#
+# The original copyright notice is as follows
+# Copyright 2021 The NetKet Authors - All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+
+
+def _treeify(f):
+    def _f(x, *args, **kwargs):
+        return jax.tree_util.tree_map(lambda y: f(y, *args, **kwargs), x)
+
+    return _f
+
+
+@_treeify
+def _unchunk(x):
+    return x.reshape((-1,) + x.shape[2:])
+
+
+@_treeify
+def _chunk(x, jac_chunk_size=None):
+    # jac_chunk_size=None -> add just a dummy chunk dimension,
+    #  same as np.expand_dims(x, 0)
+    if x.ndim == 0:
+        raise ValueError("x cannot be chunked as it has 0 dimensions.")
+    n = x.shape[0]
+    if jac_chunk_size is None:
+        jac_chunk_size = n
+
+    n_chunks, residual = divmod(n, jac_chunk_size)
+    if residual != 0:
+        raise ValueError(
+            "The first dimension of x must be divisible by jac_chunk_size."
+            + f"\n        Got x.shape={x.shape} but jac_chunk_size={jac_chunk_size}."
+        )
+    return x.reshape((n_chunks, jac_chunk_size) + x.shape[1:])
+
+
+def _jac_chunk_size(x):
+    b = set(map(lambda x: x.shape[:2], jax.tree_util.tree_leaves(x)))
+    if len(b) != 1:
+        raise ValueError(
+            "The arrays in x have inconsistent jac_chunk_size or number of chunks"
+        )
+    return b.pop()[1]
+
+
+def unchunk(x_chunked):
+    """Merge the first two axes of an array (or a pytree of arrays).
+
+    Parameters
+    ----------
+    x_chunked: an array (or pytree of arrays) of at least 2 dimensions
+
+    Returns
+    -------
+    (x, chunk_fn) : tuple
+        where x is x_chunked reshaped to (-1,)+x.shape[2:]
+        and chunk_fn is a function which restores x given x_chunked
+
+    """
+    return _unchunk(x_chunked), functools.partial(
+        _chunk, jac_chunk_size=_jac_chunk_size(x_chunked)
+    )
+
+
+def chunk(x, jac_chunk_size=None):
+    """Split an array (or a pytree of arrays) into chunks along the first axis.
+
+    Parameters
+    ----------
+        x: an array (or pytree of arrays)
+        jac_chunk_size: an integer or None (default)
+            The first axis in x must be a multiple of jac_chunk_size
+
+    Returns
+    -------
+    (x_chunked, unchunk_fn): tuple
+        - x_chunked is x reshaped to (-1, jac_chunk_size)+x.shape[1:]
+          if jac_chunk_size is None then it defaults to x.shape[0], i.e. just one chunk
+        - unchunk_fn is a function which restores x given x_chunked
+    """
+    return _chunk(x, jac_chunk_size), _unchunk
+
+
+####
+
+# The following section of this code is derived from the NetKet project
+# https://github.com/netket/netket/blob/9881c9fb217a2ac4dc9274a054bf6e6a2993c519/
+# netket/jax/_scanmap.py
+#
+# The original copyright notice is as follows
+# Copyright 2021 The NetKet Authors - All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
```

Comment:
This looks like a duplicate of the notice at the top of this file. Can we remove this one?